### PR TITLE
Fixed #27863 -- Added support for the SameSite cookie flag

### PR DESCRIPTION
**Pull request by alex**
https://code.djangoproject.com/ticket/27863
(rebased from #8380)

### CODE REVIEW

File: docs/ref/settings.txt [hunk 1]
```diff
@@ -3001,6 +3013,38 @@ session cookies from JavaScript, you're probably doing it wrong.
...
+``SESSION_COOKIE_SAMESITE``
+---------------------------
+
+Default: ``'Lax'``
+
+The value of ``SameSite`` flag on the session cookie. This flag prevents the cookie
+from being sent in cross-site requests thus preventing CSRF attacks and making some
+methods of stealing session cookie impossible.
+
+Possible values for the setting are ``'Lax'`` (default), ``'Strict'`` or ``None``.
```
Comment (timgraham) [inline docs/ref/settings.txt]: , or

File: docs/ref/settings.txt [hunk 2]
```diff
@@ -3001,6 +3013,38 @@ session cookies from JavaScript, you're probably doing it wrong.
...
+to allow any transactional pages to be linked from external sites so the ``'Strict'`` flag
+would be most appropriate here.
+
+The default ``'Lax'`` value provides a reasonable balance between security
+and usability for websites that want to maintain user's logged-in session after
+the user arrives from an external link. In the above GitHub scenario, the session cookie
+would be allowed when following a regular link from an external website while blocking
+it in CSRF-prone request methods (e.g. ``POST``).
+
+.. _SameSite: https://www.owasp.org/index.php/SameSite
```
Comment (timgraham) [inline docs/ref/settings.txt]: Link looks unused.

File: django/http/cookie.py [hunk 3]
```diff
@@ -3,6 +3,10 @@
 # For backwards compatibility in Django 2.1.
 SimpleCookie = cookies.SimpleCookie
 
+# Monkey patching the Morsel object to support SameSite attribute added in 3.8
```
Comment (timgraham) [inline django/http/cookie.py]: `# Add support for the SameSite attribute (obsolete when PY37 is unsupported).`
I think the link to the PR isn't needed.

File: django/conf/global_settings.py [hunk 4]
```diff
@@ -461,6 +461,9 @@ def gettext_noop(s):
 SESSION_COOKIE_PATH = '/'
 # Whether to use the non-RFC standard httpOnly flag (IE, FF3+, others)
 SESSION_COOKIE_HTTPONLY = True
+# Whether to set the flag restricting cookie leaks on cross-site requests
+# This can be 'Lax', 'Strict' or None to disable the flag
```
Comment (timgraham) [inline django/conf/global_settings.py]: ", or"
Please use periods.

File: docs/releases/2.1.txt [hunk 5]
```diff
@@ -279,6 +279,15 @@ Validators
...
 =====================================
 
+``SameSite`` cookies
+--------------------
+
+The cookies used for ``django.contrib.sessions``, ``django.contrib.messages``,
+and Django's CSRF protection now set the ``SameSite`` flag to ``Lax`` by
+default. This causes browsers to not send these cookies on cross-origin
+requests. If you were relying on this behavior you can set the
+``SESSION_COOKIE_SAMESITE`` or ``CSRF_COOKIE_SAMESITE`` setting to ``None``.
```
Comment (timgraham) [inline docs/releases/2.1.txt]: use links:
```
:setting:``SESSION_COOKIE_SAMESITE`
```

File: docs/releases/2.1.txt [hunk 6]
```diff
@@ -112,7 +112,7 @@ Minor features
 :mod:`django.contrib.sessions`
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-* ...
+* Added support for ``SameSite`` HTTP cookie flag for session cookies (:ticket:`27863`).
```
Comment (timgraham) [inline docs/releases/2.1.txt]: Remove ```(:ticket:`27863`)``` and add some links to the relevant documentation.

File: docs/spelling_wordlist [hunk 7]
```diff
@@ -599,6 +599,7 @@ rss
 runserver
 runtime
 safeseq
+SameSite
```
Comment (timgraham) [inline docs/spelling_wordlist]: This can be be removed as long as "SameSite" is always in backticks.

File: tests/sessions_tests/tests.py [hunk 8]
```diff
@@ -660,6 +660,21 @@ def test_httponly_session_cookie(self):
             str(response.cookies[settings.SESSION_COOKIE_NAME])
         )
 
+    @override_settings(SESSION_COOKIE_SAMESITE='Lax')
```
Comment (timgraham) [inline tests/sessions_tests/tests.py]: This is the default, it is needed? Perhaps another test for a non-default would be useful.

File: tests/sessions_tests/tests.py [hunk 9]
```diff
@@ -660,6 +660,21 @@ def test_httponly_session_cookie(self):
...
+        response = HttpResponse('Session test')
+        middleware = SessionMiddleware()
+
+        # Simulate a request the modifies the session
+        middleware.process_request(request)
+        request.session['hello'] = 'world'
+
+        # Handle the response through the middleware
+        response = middleware.process_response(request, response)
+        self.assertEqual(
```
Comment (timgraham) [inline tests/sessions_tests/tests.py]: use a single line

File: tests/httpwrappers/tests.py [hunk 10]
```diff
@@ -746,6 +746,13 @@ def test_invalid_cookies(self):
         # document.cookie parses whitespace.
         self.assertEqual(parse_cookie('  =  b  ;  ;  =  ;   c  =  ;  '), {'': 'b', 'c': ''})
 
+    def test_samesite(self):
+        c = SimpleCookie('name=value; samesite=lax; httponly')
+        # this tests both the patched version of SimpleCookie and Morsel
+        # expected output:
+        # 'Set-Cookie: name=value; HttpOnly; SameSite=lax'
+        self.assertIn('SameSite', c.output())
```
Comment (timgraham) [inline tests/httpwrappers/tests.py]: I think the comment doesn't add much value (and it'll be something we'd want to remember to remove eventually).

For the assertions, I'd use:
```
self.assertEqual(c['name']['samesite'], 'lax')
self.assertIn('SameSite=lax', c.output())
```

**Comment by alex** [comment]
@timgraham I think I got all the review feedback. Thanks!

**Comment by alex** [comment]
@kravietz I'd encourage you to apply for a Google Patch Reward for your work on this: https://www.google.com/about/appsecurity/patch-rewards/

**Comment by Flimm** [comment]
I'm excited for this release, I love this!

**Comment by alex** [comment]
Yay, I'm glad folks are excited for strong defense in depth for common web attacks!

**Comment by Gee19** [comment]
Did this not make it into 1.11?

https://github.com/django/django/pull/8380

**Comment by timgraham** [comment]
Correct, look at the release notes for the commit and you'll see it went into Django 2.1.

**Comment by Gee19** [comment]
Will it be in the next minor version (1.11.14)?

**Comment by timgraham** [comment]
Per our [supported versions policy](https://docs.djangoproject.com/en/dev/internals/release-process/#supported-versions), 1.11 is only receiving fixes for security and data loss issues.

**Comment by max-arnold** [comment]
Firefox 78.0.1 still shows the SameSite warning on the console when Django deletes the `messages` cookie:

`Set-Cookie | messages=""; expires=Thu, 01 Jan 1970 00:00:00 GMT; Max-Age=0; Path=/`

Maybe the `response.delete_cookie` method should also use the `samesite` parameter when calling `set_cookie`?

**Pull request merged**